### PR TITLE
Implemented tecdsa functionality

### PR DESCRIPTION
**Pull request by bdemann**
Added azle_try_from_js_value and azle_into_js_value for blob[] and added tests for the those
Added the tECDSA functionality and azle examples for it

### CODE REVIEW

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did [hunk 1]
```diff
@@ -0,0 +1,90 @@
+type CreateCanisterArgs = record {
```
Comment (bdemann) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did:1]: This index.did seems awfully long to me. Any idea why?

File: canisters/management.ts [hunk 2]
```diff
@@ -150,6 +150,36 @@ export type HttpResponse = {
     body: blob;
 };
 
+export type KeyID = {
```
Comment (dansteren) [inline canisters/management.ts:153]: Teeny-tiny thing... I think we want `KeyId`, not `KeyID`. Just like `Ecdsa`, I think we only capitalize the first letter of an abbreviation/acronym.

**Review by dansteren — CHANGES_REQUESTED** [review]
Overall you did an amazing job here. Just needs a few small tweaks here and there.

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did (hunk 1 above)
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did:1]: This has been addressed during the Rust rewrite

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+} from 'azle/canisters/management';
+
+export type EcdsaResult = Variant<{
+    ok: blob;
+    err: string;
+}>;
+
+export function* public_key(): Update<EcdsaResult> {
+    const caller = ic.caller().toUint8Array();
+    const public_key: CanisterResult<EcdsaPublicKeyResult> =
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:15]: I agree, public_key_result

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        yield ManagementCanister.ecdsa_public_key({
+            canister_id: null,
+            derivation_path: [caller],
+            key_id: { curve: { secp256k1: null }, name: 'dfx_test_key' }
+        });
+
+    if (!ok(public_key)) {
+        return public_key;
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:23]: You'll need to pull the error out of this, `{ err: public_key.err }`

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    const signature_result: CanisterResult<SignWithEcdsaResult> =
+        yield ManagementCanister.sign_with_ecdsa({
+            message_hash: message_hash,
+            derivation_path: [caller],
+            key_id: { curve: { secp256k1: null }, name: 'dfx_test_key' }
+        }).with_cycles(10_000_000_000n);
+
+    if (!ok(signature_result)) {
+        return signature_result;
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:40]: You'll need to pull the error out of this, `{ err: public_key.err }`

File: examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
+use ic_cdk::export::{
+    candid::CandidType,
+    serde::{Deserialize, Serialize},
+    Principal,
+};
+use ic_cdk_macros::*;
+use std::cell::RefCell;
+use std::str::FromStr;
+
+//#region Performance
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs:10]: I concur

File: examples/motoko_examples/threshold_ecdsa/canisters/motoko/main.mo [hunk 7]
```diff
@@ -0,0 +1,83 @@
+import Cycles "mo:base/ExperimentalCycles";
+import Error "mo:base/Error";
+import Prim "mo:⛔";
+import Principal "mo:base/Principal";
+
+// Define the actor
+actor Assistant {
+
+  //#region Performance
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/motoko/main.mo:9]: I concur

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+type EcdsaResult = variant { "ok": blob; "err": text };
+
+type InstallCodeMode = variant { "install": null; "reinstall": null; "upgrade": null };
+
+type CanisterStatus = variant { "running": null; "stopping": null; "stopped": null };
+
+service: {
+    "public_key": () -> (EcdsaResult);
+    "sign": (blob) -> (EcdsaResult);
+}
```
[the remark this answers is not in the document]
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/azle/index.did:90]: I concur

File: examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+#[derive(CandidType, Serialize, Debug)]
+struct SignWithECDSA {
+    pub message_hash: Vec<u8>,
+    pub derivation_path: Vec<Vec<u8>>,
+    pub key_id: EcdsaKeyId,
+}
+
+#[derive(CandidType, Deserialize, Debug)]
+struct SignWithECDSAReply {
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs:71]: Use consistent naming of Ecdsa

File: examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        .map_err(|e| format!("Failed to call ecdsa_public_key {}", e.1))?;
+
+    Ok(PublicKeyReply {
+        public_key: res.public_key,
+    })
+}
+
+#[update]
+async fn sign(message: Vec<u8>) -> Result<SignatureReply, String> {
+    assert!(message.len() == 32);
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs:113]: Do we have this check in the Azle code?

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import { blob, ok, Update, ic, Variant, CanisterResult } from 'azle';
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:1]: Do we have tests for this example?

File: canisters/management.ts [hunk 12]
```diff
@@ -150,6 +150,36 @@ export type HttpResponse = {
...
+    secp256k1: null;
+}>;
+
+export type EcdsaPublicKeyArgs = {
+    canister_id: Opt<string>;
+    derivation_path: blob[];
+    key_id: KeyID;
+};
+
+export type SignWithEcdsaArgs = {
```
Comment (lastmjs) [inline canisters/management.ts:168]: What happened to the Ethereum example you were doing, setting up the node and all of that?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        });
+
+    if (!ok(public_key)) {
+        return public_key;
+    }
+
+    return { ok: public_key.ok.public_key };
+}
+
+export function* sign(message_hash: blob): Update<EcdsaResult> {
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:29]: This functions should return results, ok, err

File: examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    const caller = ic.caller().toUint8Array();
+
+    const signature_result: CanisterResult<SignWithEcdsaResult> =
+        yield ManagementCanister.sign_with_ecdsa({
+            message_hash: message_hash,
```
Comment (lastmjs) [inline examples/motoko_examples/threshold_ecdsa/canisters/azle/index.ts:34]: You can combine `message_hash: message_hash` in to `message_hash`

File: examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs (hunk 9 above)
Comment (lastmjs) [reply examples/motoko_examples/threshold_ecdsa/canisters/rust/src/lib.rs:71]: I suppose you just copied this example over, so if that's the case then we'll just keep it as it is

**Review by lastmjs — CHANGES_REQUESTED** [review]

**Pull request merged**